mojo/core/broker_castanets.cc

```diff
@@ -206,27 +183,39 @@ bool BrokerCastanets::SyncSharedBuffer(
   if (!tcp_connection_)
     return true;
 
-  CHECK_GE(mapping.mapped_size(), offset + sync_size);
+  SyncSharedBufferImpl(mapping.guid(), static_cast<uint8_t*>(mapping.memory()),
+                       offset, sync_size, mapping.mapped_size());
+  return true;
+}
+
+void BrokerCastanets::SyncSharedBufferImpl(const base::UnguessableToken& guid,
+                                           uint8_t* memory, size_t offset,
+                                           size_t sync_size, size_t mapped_size) {
+  base::WaitableEvent* sync_wait = nullptr;
+  if (!io_thread_checker_.CalledOnValidThread()) // workaround
+    sync_wait = BeginSync(guid);
+
+  CHECK_GE(mapped_size, offset + sync_size);
   BufferSyncData* buffer_sync = nullptr;
   void* extra_data = nullptr;
   Channel::MessagePtr out_message =
       CreateBrokerMessage(BrokerMessageType::BUFFER_SYNC, 0, sync_size,
                           &buffer_sync, &extra_data);
 
-  buffer_sync->guid_high = mapping.guid().GetHighForSerialization();
-  buffer_sync->guid_low = mapping.guid().GetLowForSerialization();
+  buffer_sync->guid_high = guid.GetHighForSerialization();
+  buffer_sync->guid_low = guid.GetLowForSerialization();
   buffer_sync->offset = offset;
   buffer_sync->sync_bytes = sync_size;
-  buffer_sync->buffer_bytes = mapping.mapped_size();
-  memcpy(extra_data,
-         static_cast<uint8_t*>(mapping.memory()) + offset,
-         sync_size);
+  buffer_sync->buffer_bytes = mapped_size;
+  memcpy(extra_data, memory + offset, sync_size);
 
-  VLOG(2) << "Send Sync" << mapping.guid() << " offset: " << offset
+  VLOG(2) << "Send Sync" << guid << " offset: " << offset
           << ", sync_size: " << sync_size
           << ", buffer_size: " << buffer_sync->buffer_bytes;
   channel_->Write(std::move(out_message));
-  return true;
+
+  if (sync_wait)
+    sync_wait->Wait();
```

Comment:
Is it possible to receive ACK before Wait() call?

Reply:
No. We immediately call Wait() after Send().